usr/share/rear/lib/layout-functions.sh

```diff
@@ -982,7 +982,7 @@ function apply_layout_mappings() {
     while read original replacement junk ; do
         # Skip lines that have wrong syntax:
         test "$original" -a "$replacement" || continue
-        if grep -q "$replacement" "$file_to_migrate" ; then
+        if grep -q "'^[^#].*$replacement" "$file_to_migrate" ; then
```

Comment:
Please add a comment here that explains the reason behind that grep regexp
i.e. commented out lines that contain $replacement can be ignored and the reason
why commented out lines that contain $replacement can be ignored (see below).

The new regexp makes it no longer match at all for me
```
# replacement="_REAR123_"

# echo "# _REAR123_ foo bar" | grep "'^[^#].*$replacement" && echo y

# echo "_REAR123_ foo bar" | grep "'^[^#].*$replacement" && echo y
```
in contrast to the old regexp
```
# echo "_REAR123_ foo bar" | grep "$replacement" && echo y
_REAR123_ foo bar
y
```
so that I suggest simpler code instead of a too complicated all-in-one regexp
```
# echo "# _REAR123_ foo bar" | grep -v '^#' | grep "$replacement" && echo y

# echo "_REAR123_ foo bar" | grep -v '^#' | grep "$replacement" && echo y
_REAR123_ foo bar
y
```
i.e. code like
```
if grep -v '^#' "$file_to_migrate" | grep "$replacement" ; then
```

But I wonder why commented out lines can be ignored here at all
because in the apply_layout_mappings function the `sed` commands
in "Step1" and "Step 2" should not care about commented out lines
so that replacement hapens also in commented out lines
and then during "Step 3" no `_REAR*_` string should be left
in any line regardless whether it is commented out or not.

Simply put:
I think the `LogPrintError "Failed to apply layout mappings to ...`
that you see is correct and you would have to fix its root cause.

Reply:
1. There is a bug in my code: Somehow a single quote slipped in, which shouldn't have been there.
1.  Your test did not match because you tried a replacement at the beginning of a line. This is not supposed to happen in `disklayout.conf`, right?

If you correct my mistake and adjust the test it works:
```
replacement="_REAR123_"
echo "#disk _REAR123_ foo bar" | grep "^[^#].*$replacement" && echo y
echo "disk _REAR123_ foo bar" | grep "^[^#].*$replacement" && echo y
```

I have just tested a bit more and it seems like the case is not as obvious as I had initially thought it to be. I'll do some more testing and come up with results.

Reply:
The apply_layout_mappings function can be called with any file,
not only with $LAYOUT_FILE so a `_REAR*_` string can be anywhere.

Currently we call the apply_layout_mappings function
in the following ways, note in particular the last one:

layout/prepare/default/320_apply_mappings.sh
```
apply_layout_mappings "$LAYOUT_FILE" ...
```

finalize/GNU/Linux/260_rename_diskbyid.sh
```
OLD_ID_FILE="${VAR_DIR}/recovery/diskbyid_mappings"
...
apply_layout_mappings "$OLD_ID_FILE" ...
```

finalize/GNU/Linux/250_migrate_disk_devices_layout.sh
```
for restored_file in [b]oot/{grub.conf,menu.lst,device.map} [e]tc/grub.* [b]oot/grub/{grub.conf,menu.lst,device.map} \
                     [b]oot/grub2/{grub.conf,grub.cfg,menu.lst,device.map} \
                     [e]tc/sysconfig/grub [e]tc/sysconfig/bootloader \
                     [e]tc/lilo.conf \
                     [e]tc/yaboot.conf \
                     [e]tc/mtab [e]tc/fstab \
                     [e]tc/mtools.conf \
                     [e]tc/smartd.conf [e]tc/sysconfig/smartmontools \
                     [e]tc/sysconfig/rawdevices \
                     [e]tc/security/pam_mount.conf.xml [b]oot/efi/*/*/grub.cfg
do
    ...
    ... apply_layout_mappings "$restored_file" ...
    ...
done
```

Reply:
The original problem had appeared on bare metal systems equipped with TCG Opal 2 self-encrypting disks, so I did another round of tests on one of those. Migration mode was used as the target systems had a smaller number of disks than the original systems. With the mapping correction in place (commit fbfb9d6), the original problem (uncommented lines in disk layout containing `_REAR*_` placeholders) did not reappear.

So I have documented in my reversion commit:

The fix is unnecessary as the problem seemed to be caused
by a bug concerning the exclusion of of TCG Opal 2 self-encrypting
disks (now fixed by commit fbfb9d6).

Sorry for the annoyance.

Reply:
No need to be sorry.
I appreciate all your various usage and testing of ReaR.